File: web_ui/app/features/chat/data/datasources/chat_remote_datasource.ts
import axiosInstance from "@/app/core/interceptor";
import { ConversationDataDto } from "../dto/ConversationData.dto";
import { config } from "@/app/core/config";
import axios, { AxiosError } from "axios";
import { ConversationsDataDto } from "../dto/ConversationsData.dto";
import { Observable, ReplaySubject, shareReplay } from "rxjs";
import { ConversationsDto } from "../dto/Conversations.dto";
import { fetchEventSource } from "@microsoft/fetch-event-source";
export class ChatRemoteDatasource {
  private eventSource: AbortController | null = null;

  connectChatStream(accessToken: string): Observable<ConversationsDto> {
    return new Observable<ConversationsDto>((observer) => {
      this.eventSource = new AbortController();
      fetchEventSource(`${config.quickChatSSEService}/chat/events`, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          Accept: "text/event-stream",
        },
        signal: this.eventSource.signal,
        onopen: async (response) => {
          if (response.ok) {
            console.log("SSE connection opened.");
          } else {
            observer.error(Error(response.statusText));
          }
        },
        onmessage(event) {
          try {
            const data = JSON.parse(event.data);
            const message: ConversationsDto = {
              sender: data.sender,
              recipient: data.recipient,
              sentAt: data.sentAt,
              content: data.content,
            };
            observer.next(message);
          } catch (error) {
            observer.error("Error parsing message data");
          }
        },

        onerror(error) {
          observer.error("SSE connection error");
          console.error("SSE Error:", error);
        },
        onclose() {
          console.log("SSE connection closed.");
          observer.complete();
        },
      });
      return () => {
        console.log("SSE stream unsubscribed.");
      };
    });
  }

  async disposeChatStream(): Promise<void> {
    if (this.eventSource) {
      console.log("Disposing SSE connection...");

      // Abort the fetch event source
      this.eventSource.abort();
      this.eventSource = null;
    }
  }

  async getConversation(page: number): Promise<ConversationDataDto> {
    try {
      const rawData = {
        page: page,
      };

      const response = await axiosInstance.get<{ data: any }>(
        `${config.quickChatService}/conversation`,
        { params: rawData }
      );

      if (response.status === 200 && response.data?.data) {
        const data = response.data.data;
        return {
          currentPage: data.currentPage,
          maxPage: data.maxPage,
          conversationList: data.conversationList,
        };
      }

      throw new Error("Invalid response format");
    } catch (error) {
      throw new Error(`${error}`);
    }
  }

  async getConversations(
    conversationsId: string,
    page: number
  ): Promise<ConversationsDataDto> {
    try {
      const rawData = {
        conversationsId: conversationsId,
        page: page,
      };

      const response = await axiosInstance.get<{ data: any }>(
        `${config.quickChatService}/conversations`,
        { params: rawData }
      );

      if (response.status === 200 && response.data?.data) {
        const data = response.data.data;
        return {
          currentPage: data.currentPage,
          maxPage: data.maxPage,
          conversationsList: data.conversationsList,
        };
      }

      throw new Error("Invalid response format");
    } catch (error) {
      throw new Error(`${error}`);
    }
  }

  async sendMessage(recipient: string, content: string): Promise<boolean> {
    try {
      const rawData = {
        recipient: recipient,
        content: content,
      };

      const response = await axiosInstance.post<{ data: any }>(
        `${config.quickChatService}/chat/send`,
        rawData
      );

      // trigger sse
      await axiosInstance.post<{ data: any }>(
        `${config.quickChatSSEService}/chat/send`,
        rawData
      );

      if (response.status === 201) {
        return true;
      }

      throw new Error("Invalid response format");
    } catch (error) {
      throw new Error(`${error}`);
    }
  }
}

File: mobile_ui/quickchat_ui/lib/features/chat/data/datasources/chat_remote_datasource.dart
import 'dart:convert';

import 'package:dartz/dartz.dart';
import 'package:dio/dio.dart';
import 'package:injectable/injectable.dart';
import 'package:quickchat_ui/config.dart';
import 'package:quickchat_ui/core/logic/server_exeption.dart';
import 'package:quickchat_ui/core/sse/quick_chat_sse.dart';
import 'package:quickchat_ui/features/chat/data/dto/ConversationData.dto.dart';
import 'package:quickchat_ui/features/chat/data/dto/Conversations.dto.dart';
import 'package:quickchat_ui/features/chat/data/dto/ConversationsData.dto.dart';

abstract class ChatRemoteDatasource {
  Stream<dynamic> connectChatStream(
    String accessToken,
  );
  void disposeChatStream();

  Future<ConversationDataDto> getConversation({required int page});
  Future<ConversationsDataDto> getConversations({
    required String conversationsId,
    required int page,
  });
  Future<Unit> sendChat({
    required String recipient,
    required String content,
  });
}

@LazySingleton(as: ChatRemoteDatasource)
class ChatRemoteDatasourceImpl implements ChatRemoteDatasource {
  ChatRemoteDatasourceImpl({
    required this.dio,
    required this.config,
    required this.quickChatSse,
  });

  final Dio dio;
  final Config config;
  final QuickChatSse quickChatSse;

  @override
  Stream connectChatStream(String accessToken) {
    final Stream<dynamic> stream = quickChatSse.connectChatStream(accessToken);

    return stream.map((dynamic event) {
      if (event is Map<String, dynamic>) {
        return ConversationsDto.fromJson(event);
      }

      return null;
    }).where((ConversationsDto? event) => event != null);
  }

  @override
  void disposeChatStream() {
    quickChatSse.disposeChatStream();
  }

  @override
  Future<ConversationDataDto> getConversation({required int page}) async {
    late Response<dynamic> response;
    late ConversationDataDto resultData;

    try {
      final Map<String, dynamic> rawData = <String, dynamic>{
        'page': page,
      };
      // Fetch conversation data
      response = await dio.get(
        '${config.quickChatService}/conversation',
        data: jsonEncode(rawData),
      );

      final Map<String, dynamic> result = response.data['data'];

      resultData = ConversationDataDto.fromJson(result);
    } on DioException catch (e) {
      throw checkErrResponse(e.response);
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 200) {
      return resultData;
    } else {
      throw checkErrResponse(response);
    }
  }

  @override
  Future<ConversationsDataDto> getConversations({
    required String conversationsId,
    required int page,
  }) async {
    late Response<dynamic> response;
    late ConversationsDataDto resultData;

    try {
      final Map<String, dynamic> rawData = <String, dynamic>{
        'conversationsId': conversationsId,
        'page': page,
      };
      // Fetch conversations data
      response = await dio.get(
        '${config.quickChatService}/conversations',
        data: jsonEncode(rawData),
      );
      final Map<String, dynamic> result = response.data['data'];

      resultData = ConversationsDataDto.fromJson(result);
    } on DioException catch (e) {
      throw checkErrResponse(e.response);
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 200) {
      return resultData;
    } else {
      throw checkErrResponse(response);
    }
  }

  @override
  Future<Unit> sendChat({
    required String recipient,
    required String content,
  }) async {
    late Response<dynamic> response;

    try {
      // Prepare the data
      final Map<String, dynamic> rawData = <String, dynamic>{
        'recipient': recipient,
        'content': content,
      };

      // Send the chat message
      response = await dio.post(
        '${config.quickChatService}/chat/send',
        data: jsonEncode(rawData),
      );
      // Trigger the sse
      dio.post('${config.quickChatSSEService}/chat/send',
          data: jsonEncode(rawData));
    } on DioException catch (e) {
      throw checkErrResponse(e.response);
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 201) {
      return unit;
    } else {
      throw checkErrResponse(response);
    }
  }
}
